### ISSUE TITLE
Spawned child processes print to Terminal instead of Jupyter Cell Output

### ISSUE THREAD

**Issue by LarsHill** [jupyterlab/jupyterlab#9668]
## Description

When spawning a set of child processes using pythons `multiprocessing` library, `print` statements executed from within a child process get send to the Terminal session that launched the Jupyter client instead of showing up as usual in the cell output of the Jupyter UI. `print` statements executed from the parent process behave normally and their output shows up in the Jupyter UI below the executed cell.

## Reproduce

Here is a minimal working example that shows this behaviour:
**Note**: We have to import the target function from an external script, since Jupyter throws attribute errors when defining it within the Notebook. See this [Stackoverflow](https://stackoverflow.com/questions/41385708/multiprocessing-example-giving-attributeerror).

File: `target.py`
```python
def x_squared(x: int):
    print(x**2)
```

File: `min_example.ipynb`
```python
import multiprocessing as mp

from target import x_squared

processes=[mp.Process(target=x_squared, args=(x,)) for x in range(1,4)]

for p in processes:
    p.start()

for p in processes:
    p.join()
```

When executing the above code from a single cell in the notebook the prints of the target are not shown as output of the cell. Instead, they are send to the Terminal, which launched Jupyter (as described above).

## Expected behavior

I would expect that `print` statements from child processes also get send to the Jupyter Frontend instead of showing up in the Terminal session that launched Jupyter. At first, I wasn't aware of this behaviour and thought something is wrong with my code since no results and prints where shown. Only accidentally I found out that everything was displayed in the Terminal session, which I think is not clear to users and should be changed.

Is there any way of forcing Jupyter to also print results from child processes as cell output?

## Context

- Operating System and version: Mac OS 10.12.6
- Browser and version: Google Chrome 87.0.4280.88
- JupyterLab version: 2.2.9

**Comment by jasongrout** [comment]
I think this is a more generic issue with launching processes in python and redirecting their stdout to the same stdout file descriptor of the parent process, not an issue with JupyterLab per se. You can check that you have the same problem in the classic Jupyter Notebook.

Concepts to search for include: 
* redirecting stdout or stderr when spawning a process
* you might be able to use the subprocess module, which provides facilities for capturing stdout from subprocesses
* you might need to set up a more general message passing mechanism between the parent and child process to get data from the child process. I think the multiprocessing library has data structures for passing information
* Perhaps using the threading library helps? (not sure...)

The relevant thing here about JupyterLab/Jupyter Notebook is that they replace the stdout file descriptor of the main kernel process with something that will capture output and send it to the notebook. If you do not spawn the child process with that file descriptor or somehow correct for that, the child process will simply use the standard stdout file descriptor.

**Comment by jasongrout** [comment]
(For example, at a quick glance, this looks possibly helpful: https://stackoverflow.com/questions/7714868/python-multiprocessing-how-can-i-reliably-redirect-stdout-from-a-child-process)

**Comment by LarsHill** [comment]
Thanks for your extensive reply!
In the meantime we also did some deeper research on the topic and came to a similar conclusion. We basically circumvented the issue by going with the solution you outlined in your third bullet point:

> message passing mechanism between the parent and child process

In more detail, we applied the approach described in the logging coocbook [logging-to-a-single-file-from-multiple-processes](https://docs.python.org/3/howto/logging-cookbook.html#logging-to-a-single-file-from-multiple-processes), using the `QueueHandler` class from the `logging` library to forward all child process logs to a separate `QueueListener` Thread in the main process. Then the logs get handled by a configured logger in the main process. 
With that approach Jupyter also shows all the output logs, since they are now coming from the main process itself.

So I think this issue can be closed :)
